encfs/MACFileIO.cpp

```diff
@@ -137,84 +141,130 @@ off_t MACFileIO::getSize() const {
+          int test = mac & 0xff;
+          int stored = tmp.data[i];
 
-      // Constant time comparision to prevent timing attacks
-      unsigned char fail = 0;
-      for (int i = 0; i < macBytes; ++i, mac >>= 8) {
-        int test = mac & 0xff;
-        int stored = tmp.data[i];
+          fail |= (test ^ stored);
+        }
+
+          if (fail > 0) {
+            // uh oh..
+            long blockNum = req.offset / bs;
+            RLOG(WARNING) << "MAC comparison failure in block " << blockNum;
+            if (!warnOnly) {
+              MemoryPool::release(mb);
+              throw Error(_("MAC comparison failure, refusing to read"));
+            }
+          }
+        }
 
-        fail |= (test ^ stored);
+        // now copy the data to the output buffer
+        readSize -= headerSize;
+        memcpy(req.data, tmp.data + headerSize, readSize);
+      } else {
+        VLOG(1) << "readSize " << readSize << " at offset " << req.offset;
+        if (readSize > 0) readSize = 0;
       }
 
-      if (fail > 0) {
-        // uh oh..
-        long blockNum = req.offset / bs;
-        RLOG(WARNING) << "MAC comparison failure in block " << blockNum;
-        if (!warnOnly) {
-          MemoryPool::release(mb);
-          throw Error(_("MAC comparison failure, refusing to read"));
+      MemoryPool::release(mb);
+
+      return readSize;
+
+  } else {	  //reverse = true
+
+      if (randBytes > 0) {
+        RLOG(ERROR) << "MAC rand bytes not yet implemented in reverse mode";
+        return -EPERM;
+      }
+
+      MemBlock mb = MemoryPool::allocate(blockSize());
+
+      IORequest tmp;
+      tmp.offset = locWithoutHeader(req.offset, blockSize(), macBytes);
+      tmp.data = mb.data + macBytes;
+      tmp.dataLen = req.dataLen - macBytes;
+
+      // get the data from the base FileIO layer
+      ssize_t readSize = base->read(tmp);
+      if (readSize <= 0) {
+        MemoryPool::release(mb);
+        return readSize;
+      }
+
+      if (macBytes > 0) {
+        uint64_t mac = cipher->MAC_64(tmp.data, readSize, key);
+
+        //point to the beginning to write the header
+        tmp.data = mb.data;
+
+        for (int i = 0; i < macBytes; ++i) {
+          tmp.data[i] = mac & 0xff;
+          mac >>= 8;
         }
       }
-    }
 
-    // now copy the data to the output buffer
-    readSize -= headerSize;
-    memcpy(req.data, tmp.data + headerSize, readSize);
-  } else {
-    VLOG(1) << "readSize " << readSize << " at offset " << req.offset;
-    if (readSize > 0) {
-      readSize = 0;
-    }
-  }
+      // now copy the data to the output buffer
+      memcpy(req.data, tmp.data, readSize + macBytes);
+      MemoryPool::release(mb);
 
-  MemoryPool::release(mb);
+      return readSize + macBytes;
 
-  return readSize;
+    }
 }
 
 bool MACFileIO::writeOneBlock(const IORequest &req) {
+  if (reverse) {
+    RLOG(ERROR) << "MAC write not implemented in reverse mode";
+    return false;
```

Comment:
As now we have `--reversewrite`,  perhaps we would have to do a MAC comparison before writing the block, as in `readOneBlock()` above ?